pylabrobot/plate_reading/biotek_backend.py

```diff
@@ -0,0 +1,250 @@
+
+    cmd = f"{614220 + int(1000*focal_height)}\x03".encode()
+    await self.send_command(cmd, purge=False)
+
+    resp = await self.send_command("y", wait_for_char=b"\x06")
+    assert resp == b"\x06"
+    await self.send_command(b"08120112207434014351135308559127881772\x03", purge=False)
+
+    resp = await self.send_command("D", wait_for_char=b"\x06")
+    assert resp == b"\x06"
+    excitation_wavelength_str = str(excitation_wavelength).zfill(4)
+    emission_wavelength_str = str(emission_wavelength).zfill(4)
+    cmd = (f"008401010108120001200100001100100000135000100200200{excitation_wavelength_str}000"
+      f"{emission_wavelength_str}000000000000000000210011").encode()
+    checksum = str((sum(cmd)+7) % 100).encode() # don't know why +7
+    cmd = cmd + checksum + b"\x03"
+    await self.send_command(cmd, purge=False)
+
+    resp1 = await self.send_command("O", wait_for_char=b"\x06")
+    assert resp1 == b"\x06"
+    resp2 = await self._read_until(b"\x03")
+    assert resp2 == b"0000\x03"
+
+    # read data
+    body = await self._read_until(b"\x03", timeout=60*2)
+    assert body is not None
+    return self._parse_body(body)
+
+  async def _abort(self) -> None:
+    await self.send_command("x", wait_for_char=None)
+
+  class ShakeType(enum.Enum):
+    LINEAR = "linear"
+    ORBITAL = "orbital"
+
+  async def shake(self, shake_type: ShakeType) -> None:
+    """ Warning: the duration for shaknig has to be specified on the machine, and the maximum is
+    16 minutes and 39 seconds. As a hack, we start shaking for the maximum duration every time
+    as long as stop is not called. """
+    max_duration = 16*60 + 39 # 16m39s
+
+    async def shake_maximal_duration():
+      resp = await self.send_command("y", wait_for_char=b"\x06")
+      assert resp == b"\x06"
+      await self.send_command(b"08120112207434014351135308559127881422\x03", purge=False)
+
+      resp = await self.send_command("D", wait_for_char=b"\x06")
+      assert resp == b"\x06"
+      shake_type_bit = "0" if shake_type == self.ShakeType.LINEAR else "1"
+
+      duration = str(max_duration).zfill(3)
+      cmd = f"0033010101010100002000000013{duration}{shake_type_bit}301".encode()
+      checksum = str((sum(cmd)+73) % 100).encode() # don't know why +73
+      cmd = cmd + checksum + b"\x03"
+      await self.send_command(cmd, purge=False)
+
+      resp = await self.send_command("O", wait_for_char=b"\x06")
+      assert resp == b"\x06"
+      resp = await self._read_until(b"\x03")
+      assert resp == b"0000\x03"
+
+    def shake_continuous(loop):
+      while self._shaking:
+        asyncio.run_coroutine_threadsafe(shake_maximal_duration(), loop)
+
+        # short sleep allows = frequent checks for fast stopping
+        seconds_since_start = 0
+        while seconds_since_start < max_duration and self._shaking:
+          time.sleep(0.5)
+          seconds_since_start += 1
+
+    self._shaking = True
+    self._shaking_thread = threading.Thread(target=shake_continuous,
```

Comment:
is there a reason why you are using the threading library instead of asyncio? ChatGPT suggests this works:

```
import asyncio

async def shake(self, shake_type: ShakeType) -> None:
    """ Warning: the duration for shaking has to be specified on the machine, and the maximum is
    16 minutes and 39 seconds. As a hack, we start shaking for the maximum duration every time
    as long as stop is not called. """
    max_duration = 16*60 + 39  # 16m39s

    async def shake_maximal_duration():
        resp = await self.send_command("y", wait_for_char=b"\x06")
        assert resp == b"\x06"
        await self.send_command(b"08120112207434014351135308559127881422\x03", purge=False)

        resp = await self.send_command("D", wait_for_char=b"\x06")
        assert resp == b"\x06"
        shake_type_bit = "0" if shake_type == self.ShakeType.LINEAR else "1"

        duration = str(max_duration).zfill(3)
        cmd = f"0033010101010100002000000013{duration}{shake_type_bit}301".encode()
        checksum = str((sum(cmd) + 73) % 100).encode()  # don't know why +73
        cmd = cmd + checksum + b"\x03"
        await self.send_command(cmd, purge=False)

        resp = await self.send_command("O", wait_for_char=b"\x06")
        assert resp == b"\x06"
        resp = await self._read_until(b"\x03")
        assert resp == b"0000\x03"

    async def shake_continuous():
        while self._shaking:
            await shake_maximal_duration()

            # short sleep allows frequent checks for fast stopping
            seconds_since_start = 0
            while seconds_since_start < max_duration and self._shaking:
                await asyncio.sleep(0.5)
                seconds_since_start += 0.5

    self._shaking = True
    self._shaking_task = asyncio.create_task(shake_continuous())

async def stop_shaking(self) -> None:
    await self._abort()
    if self._shaking:
        self._shaking = False
        if self._shaking_task is not None:
            self._shaking_task.cancel()
            try:
                await self._shaking_task
            except asyncio.CancelledError:
                pass
            self._shaking_task = None
```